include/deal.II/base/tensor_polynomials_base.h

```diff
@@ -0,0 +1,180 @@
+ *   <li> <tt>PolynomialsNedelec</tt>
+ *   <li> <tt>PolynomialsRaviartThomas</tt>
+ *   <li> <tt>PolynomialsRT_Bubbles</tt>
+ * </ul>
+ *
+ * @ingroup Polynomials
+ * @author Graham Harper
+ * @date 2019
+ */
+template <int dim>
+class TensorPolynomialsBase
+{
+public:
+  /**
+   * Constructor. This takes the degree of the space, @p deg from the finite element
+   * class, and @p n, the number of polynomials for the space.
+   */
+  TensorPolynomialsBase(const unsigned int deg, const unsigned int n_polynomials);
+
+  /**
+   * Move constructor.
+   */
+  TensorPolynomialsBase(TensorPolynomialsBase<dim> &&) = default; // NOLINT
+
+  /**
+   * Copy constructor.
+   */
+  TensorPolynomialsBase(const TensorPolynomialsBase<dim> &) = default;
+
+  /**
+   * Virtual destructor. Makes sure that pointers to this class are deleted
+   * properly.
+   */
+  virtual ~TensorPolynomialsBase() = default;
+
+  /**
+   * Compute the value and the derivatives of the polynomials at
+   * @p unit_point.
+   *
+   * The size of the vectors must either be zero or equal <tt>n()</tt>.  In
+   * the first case, the function will not compute these values.
+   *
+   * If you need values or derivatives of all polynomials then use this
+   * function, rather than using any of the <tt>compute_value</tt>,
+   * <tt>compute_grad</tt> or <tt>compute_grad_grad</tt> functions, see below,
+   * in a loop over all tensor product polynomials.
+   */
+  virtual void
+  compute(const Point<dim> &           unit_point,
+          std::vector<Tensor<1, dim>> &values,
+          std::vector<Tensor<2, dim>> &grads,
+          std::vector<Tensor<3, dim>> &grad_grads,
+          std::vector<Tensor<4, dim>> &third_derivatives,
+          std::vector<Tensor<5, dim>> &fourth_derivatives) const = 0;
+
+  /**
+   * Return the number of polynomials.
+   */
+  unsigned int
+  n() const;
+
+  /**
+   * Return the highest polynomial degree of polynomials represented by this
+   * class. A derived class may override this if its value is different from
+   * @p my_degree.
+   */
+  unsigned int
+  degree() const;
+
+  /**
+   * A sort of virtual copy constructor, this function returns a copy of
+   * the polynomial space object. Derived classes need to override the function
+   * here in this base class and return an object of the same type as the
+   * derived class.
+   *
+   * Some places in the library, for example the constructors of FE_PolyTensor,
+   * need to make copies of polynomial spaces without knowing their exact type.
+   * They do so through this function.
+   */
+  virtual std::unique_ptr<TensorPolynomialsBase<dim>>
+  clone() const = 0;
+
+  /**
+   * Return the name of the space.
+   */
+  virtual std::string
+  name() const = 0;
+
+  /**
+   * Return the number of polynomials in the space based on the polynomial space
+   * degree @p deg. This must be overridden by the derived class.
+   */
+  virtual unsigned int
+  compute_n_pols(const unsigned int deg) const = 0;
```

Comment:
Can't you get rid of this function now? Wouldn't derived classes simply return the same thing that the `n()` function above already returns?

Reply:
Yes I can remove that. Is there any way that we can require derived classes to implement a `static unsigned int compute_n_pols(const unsigned int deg)`?

Reply:
No, the only way for doing that are pure `virtual` functions, but `virtual` functions can't be `static`. You have to be explicit about the class anyway if you want to use this function for initializing a `FiniteElement`.